---
title: "Compare bisulfite calls vs EMSEQ vs MEGALODON"
author: "P_Terzian"
date: "`r Sys.Date()`"
output:
  prettydoc::html_pretty:
  theme: architect
highlight: github
---

Commande megalodon
```
megalodon fetch_CM003781.1_fast5/ \
    --outputs basecalls mods \
    --reference CM003781.1.fa \
    --mod-motif Z CG 0 --devices 0 --processes 20 \
    --verbose-read-progress 3 \
    --guppy-server-path /usr/bin/guppy_basecall_server \
    --overwrite
```


```{r, include=FALSE}
library(data.table)
library(Biostrings)
library(ggplot2)
library(DT)
library(reshape2)
# library(Rgb)
# library(cowplot)
# library(rlist)
# library(scales)

#source('~/Documents/axis2-epigenectics/script_BSseq/BS_process_functions.R')
#source('plot_functions_aroundTSS.R')

```

```{r}

getCpGcontextInref <- function(reference, patternToFind = "CG"){
  
  list_CpGTable <- lapply(seq_along(reference), function(x){
  
      currentChrmName <- strsplit(names(reference[x]), " ")[[1]][1]
  
        chrm_pattern_table <- as.data.frame(matchPattern(pattern = patternToFind, subject = reference[[x]])@ranges)
  
      chrm_pattern_table$chrm <-rep(currentChrmName, nrow(chrm_pattern_table))
      
      return(chrm_pattern_table)

  })
AllChrmCpGTable <- do.call(rbind, list_CpGTable)
  

}

```


### Load genome reference
```{r}

sus_scrofa <- readDNAStringSet("/work/project/seqoccin/epigenetique/sus_scrofa/GCA_000003025.6_Sscrofa11.1_genomic.fa")
#chrm1_japonica$`CM003781.1 Coturnix japonica isolate 7356 chromosome 1, whole genome shotgun sequence`[CM003781[strand == "+"]$start]

names_chrm <- sapply(seq_along(sus_scrofa), function(x){
  
      currentChrmName <- strsplit(names(sus_scrofa[x]), " ")[[1]][1]
      
})


sus_scrofa_onlychrm <- sus_scrofa[grep(x = names_chrm, "^CM")]

rm(sus_scrofa)
gc()

CG_pos_fullreference <- getCpGcontextInref(sus_scrofa_onlychrm)
setDT(CG_pos_fullreference)
setkey(CG_pos_fullreference, chrm, start)

### Compute numbers of sites per chromosome
nb_CG_per_chrm <- as.data.table(table(CG_pos_fullreference$chrm))

colnames(nb_CG_per_chrm) <- c("chrm", "position")

nb_CG_per_chrm$type <- "reference"
```


### LOAD megalodon both runs results
```{r}
mega_results_run1 <- fread(file = "/work/project/seqoccin/epigenetique/sus_scrofa/Reference_father/megalodon_calling/sus_scrofa_ref_PAD93746/modified_bases.5mC.bed", stringsAsFactors = FALSE)
colnames(mega_results_run1) <- c("chrm", "start", "stop", "score", "coverage", "strand", "start2", "stop2", "Idonknow", "coverage", "fraction")
hist(mega_results_run1$fraction)
setkey(mega_results_run1, chrm, start)
mega_results_run1 <- mega_results_run1[strand == "+", start := start + 1]
mega_merged1 <- mega_results_run1[, .(coverage = sum(coverage), fraction = sum(fraction) / length(fraction)), by = .(chrm, start)]
setkey(mega_merged1, chrm, start)
mega_merged1 <- mega_merged1[coverage > 5 & coverage < 100]
nrow(mega_merged1)

mega_results_run2 <- fread(file = "/work/project/seqoccin/epigenetique/sus_scrofa/Reference_father/megalodon_calling/sus_scrofa_ref_PAE20806/modified_bases.5mC.bed", stringsAsFactors = FALSE)
colnames(mega_results_run2) <- c("chrm", "start", "stop", "score", "coverage", "strand", "start2", "stop2", "Idonknow", "coverage", "fraction")
setkey(mega_results_run2, chrm, start)
mega_results_run2 <- mega_results_run2[strand == "+", start := start + 1]
mega_merged2 <- mega_results_run2[, .(coverage = sum(coverage), fraction = sum(fraction) / length(fraction)), by = .(chrm, start)]
setkey(mega_merged2, chrm, start)
mega_merged2 <- mega_merged2[coverage > 5 & coverage < 100]
nrow(mega_merged2)

setkey(mega_merged1, chrm, start)
setkey(mega_merged2, chrm, start)


rm(list = c("mega_results_run1", "mega_results_run2"))


```




### Load bismark cytosine_report POOLED
```{r}
pathToFiles = "/work2/project/seqoccin/epigenetique/CelineV/Project_A2P1b-PlaGe-Porc.1470/Pool_run_bisulfite_porc_ref/MethylSeq-nfcore/Sscrofa_per-lane_param-zymo/results/bismark_methylation_calls/stranded_CpG_report/"

list_files <- list.files(path = pathToFiles)

wgbsfiletype <- c('wgbs1', 'wgbs2', 'wgbs3', 'novaseq')

study_DT <- lapply(1:length(list_files), function(x){
    
    bedcoverageFile           <- fread(file = paste(pathToFiles, list_files[x], sep=""))
    colnames(bedcoverageFile) <- c("chrm", "start", "strand", "Cyt", "Thy", "di_context", "tri_context")
    bedcoverageFile$type <- wgbsfiletype[x]
    return(bedcoverageFile)
  })

study_DT    <- do.call(rbind, study_DT)

study_DT <- study_DT[strand == "-", start := start - 1]

merged_study_DT_typed <- study_DT[, .(Cyt = sum(Cyt), Thy = sum(Thy)), by = .(chrm, start, type)]

merged_study_DT_typed <- merged_study_DT_typed[Cyt != 0 | Thy != 0]


merged_study_DT_typed[, fraction := Cyt / (Cyt + Thy)]
merged_study_DT_typed[, cov := Cyt + Thy ]


rm(study_DT)
gc()


wgbs1 <- merged_study_DT_typed[cov > 5 & cov < 100 & type == "wgbs1"][]
wgbs2 <- merged_study_DT_typed[cov > 5 & cov < 100 & type == "wgbs2"][]
wgbs3 <- merged_study_DT_typed[cov > 5 & cov < 100 & type == "wgbs3"][]
novaseq <- merged_study_DT_typed[cov > 5 & cov < 100 & type == "novaseq"][]


nrow(wgbs1)
nrow(wgbs2)
nrow(wgbs3)
nrow(novaseq)

setkey(wgbs1, chrm, start)
setkey(wgbs2, chrm, start)
setkey(wgbs3, chrm, start)
setkey(novaseq)

rm(merged_study_DT_typed)
gc()

```




### Compare megalodon run with themself and BS and Emseq
```{r}

#Megapooled vs BSpooled
#Mega1 vs Mega2
mega1vmega2 <- mega_merged1[mega_merged2, nomatch=0]

nbrow_ont <- nrow(mega1vmega2)

##wgbs1
wgbs1vwgbs2     <- wgbs1[wgbs2, nomatch=0]
nbrow_wgbs1vwgbs2 <- nrow(wgbs1vwgbs2)
wgbs1vwgbs3     <- wgbs1[wgbs3, nomatch=0]
nbrow_wgbs1vwgbs3 <- nrow(wgbs1vwgbs3)
wgbs2vwgbs3     <- wgbs2[wgbs3, nomatch=0]
nbrow_wgbs2vwgbs3 <- nrow(wgbs2vwgbs3)


wgbs1vENovaseq  <- wgbs1[novaseq, nomatch=0]
wgbs2vENovaseq  <- wgbs2[novaseq, nomatch=0]
wgbs3vENovaseq  <- wgbs3[novaseq, nomatch=0]
nbrow_wgbs1vENovaseq <- nrow(wgbs1vENovaseq)
nbrow_wgbs2vENovaseq <- nrow(wgbs2vwgbs3)
nbrow_wgbs3vENovaseq <- nrow(wgbs3vENovaseq)

cormega1vmega2    <- cor(mega1vmega2$fraction, mega1vmega2$i.fraction)



corwgbs1vwgbs2    <- cor(wgbs1vwgbs2$fraction, wgbs1vwgbs2$i.fraction)
corwgbs1vwgbs3    <- cor(wgbs1vwgbs3$fraction, wgbs1vwgbs3$i.fraction)
corwgbs2vwgbs3    <- cor(wgbs2vwgbs3$fraction, wgbs2vwgbs3$i.fraction)

corwgbs1vENovaseq <- cor(wgbs1vENovaseq$fraction,wgbs1vENovaseq$i.fraction)
corwgbs2vENovaseq <- cor(wgbs2vENovaseq$fraction,wgbs2vENovaseq$i.fraction)
corwgbs3vENovaseq <- cor(wgbs3vENovaseq$fraction,wgbs3vENovaseq$i.fraction)

```

## ONT

```{r}

ggplot(mega1vmega2, aes(x=fraction,y=i.fraction)) + 
                    stat_binhex() +
                    scale_fill_gradientn(colours = c("#59539f", "#73c3a2", "#f5f9b5", "#b9304c"), trans="log10") +
                    #scale_y_log10() +
                    theme_bw() +
                    labs(title = paste("Megalodon run2 vs run3 at: ",nbrow_ont, " CpGi. Correlation :",  round(cormega1vmega2, digits=3))) +
                    xlab("Run 1 fractions") +
                    ylab("Run 2 fractions")


```


## Bisulfite

```{r}

ggplot(wgbs1vwgbs2, aes(x=fraction,y=i.fraction)) + 
                    stat_binhex() +
                    scale_fill_gradientn(colours = c("#59539f", "#73c3a2", "#f5f9b5", "#b9304c"), trans="log10") +
                    #scale_y_log10() +
                    theme_bw() +
                    labs(title = paste("BS_hiseq run1 vs BS_hiseq run2 at: ",nbrow_wgbs1vwgbs2, " CpGi. Correlation :",  round(corwgbs1vwgbs2, digits=3))) +
                    xlab("Run 1 hiseq fractions") +
                    ylab("Run 2 hiseq fractions")


ggplot(wgbs1vwgbs3, aes(x=fraction,y=i.fraction)) + 
                    stat_binhex() +
                    scale_fill_gradientn(colours = c("#59539f", "#73c3a2", "#f5f9b5", "#b9304c"), trans="log10") +
                    #scale_y_log10() +
                    theme_bw() +
                    labs(title = paste("BS_hiseq run1 vs BS_hiseq run3 at: ",nbrow_wgbs1vwgbs3, " CpGi. Correlation :",  round(corwgbs1vwgbs3, digits=3))) +
                    xlab("Run 1 hiseq fractions") +
                    ylab("Run 3 hiseq fractions")


ggplot(wgbs2vwgbs3, aes(x=fraction,y=i.fraction)) + 
                    stat_binhex() +
                    scale_fill_gradientn(colours = c("#59539f", "#73c3a2", "#f5f9b5", "#b9304c"), trans="log10") +
                    #scale_y_log10() +
                    theme_bw() +
                    labs(title = paste("BS_hiseq run2 vs BS_hiseq run3 at: ",nbrow_wgbs2vwgbs3, " CpGi. Correlation :",  round(corwgbs2vwgbs3, digits=3))) +
                    xlab("Run 2 hiseq fractions") +
                    ylab("Run 3 hiseq fractions")



ggplot(wgbs1vENovaseq, aes(x=fraction,y=i.fraction)) + 
                    stat_binhex() +
                    scale_fill_gradientn(colours = c("#59539f", "#73c3a2", "#f5f9b5", "#b9304c"), trans="log10") +
                    #scale_y_log10() +
                    theme_bw() +
                    labs(title = paste("BS_hiseq run1 vs BS_novaseq at: ",nbrow_wgbs1vENovaseq, " CpGi. Correlation :",  round(corwgbs1vENovaseq, digits=3))) +
                    xlab("Run 1 hiseq fractions") +
                    ylab("Run novaseq fractions")


ggplot(wgbs2vENovaseq, aes(x=fraction,y=i.fraction)) + 
                    stat_binhex() +
                    scale_fill_gradientn(colours = c("#59539f", "#73c3a2", "#f5f9b5", "#b9304c"), trans="log10") +
                    #scale_y_log10() +
                    theme_bw() +
                    labs(title = paste("BS_hiseq run2 vs BS_novaseq at: ",nbrow_wgbs2vENovaseq, " CpGi. Correlation :",  round(corwgbs2vENovaseq, digits=3))) +
                    xlab("Run 2 hiseq fractions") +
                    ylab("Run novaseq fractions")


ggplot(wgbs3vENovaseq, aes(x=fraction,y=i.fraction)) + 
                    stat_binhex() +
                    scale_fill_gradientn(colours = c("#59539f", "#73c3a2", "#f5f9b5", "#b9304c"), trans="log10") +
                    #scale_y_log10() +
                    theme_bw() +
                    labs(title = paste("BS_hiseq run3 vs BS_novaseq at: ",nbrow_wgbs3vENovaseq, " CpGi. Correlation :",  round(corwgbs3vENovaseq, digits=3))) +
                    xlab("Run 3 hiseq fractions") +
                    ylab("Run novaseq fractions")

```

## EM-seq

