# SeqOccIn - A Bos taurus sequencing methods benchmark for assembly, haplotyping, and variant calling

Following the Genome in a Bottle project example, we sequenced one Charolais heifer with different technologies: Illumina paired-end, Oxford Nanopore, Pacific Biosciences (HiFi and CLR), 10X Genomics linked-reads, and Hi-C. In order to generate haplotypic assemblies, we also sequenced both parents with short reads. From these data, we built two haplotyped trio high quality reference genomes and a consensus assembly, using up-to-date software packages. The assemblies obtained using PacBio HiFi reaches a size of 3.2 Gb, which is significantly larger than the 2.7 Gb ARS-UCD1.2 reference. The BUSCO score of the consensus assembly reaches a completeness of 95.8%, among highly conserved mammal genes. We also identified 35,866 structural variants larger than 50 base pairs. This assembly is a contribution to the bovine pangenome for the “Charolais” breed.  These datasets will prove to be useful resources enabling the community to gain additional insight on sequencing technologies for applications such as SNP, indel or structural variant calling, and de novo assembly.

The publication about these data is available here : xxxxxxxxxxxxxxx.publication.com

All the data produced from this projet are available under the the BioProject PRJEB55064 at EMBL/EBI ENA.

Our reference assembly is also available at EMBLI/EBI ENA under the same BioProject with accession ERZ14208621.

Others produced assemblies are available under the DOI:10.57745/73DGRC.

Our phased assemblies are available under the DOI:10.57745/40STPR

Our Short Variation detections are available under the DOI:10.57745/H4QHB1
