# SeqOccIn

The SeqOccIn project aims to compare the potential of different front-end sequencing technologies (Illumina, Pacific Biosciences, Oxford Nanopore) and methodologies in three scientific areas: genome assembly and structural variant identification (4263 samples), methylation mark detection (224 samples), and metagenomics (436 samples).

Long and short fragment sequencing data were generated for various agronomically important animal and plant species: including cattle, goat, sheep, pig, quail, and corn. In total, more than 4900 samples were sequenced and used in the following applications:  
- i) genome assembly (Long reads, Hi-C and 10x linked reads),  
- ii) punctual and structural variations analyses (including low pass sequencing),  
- iii) methylation mark analyses from both native DNA (Oxford Nanopore) and converted DNA based approaches (EMseq, Bisulfite sequencing),  
- iv) metabarcoding analyses (long read 16S and operon sequencing), and whole genome metagenomics analyses (PacBio HiFi).  

The INRAE "Sequencing Occitanie Innovation" (SeqOccIn) project, jointly supported by Genotoul GeT-PlaGe and Bioinfo platforms, was selected as part of AP01 axis, "Stimulate innovation", of the Occitanie region call for projects entitled "Regional Platforms for Research and Innovation", from the Operational program ERDF-FSE MIDI-PYRENEES ET GARONNE 2014-2020. 
