---
title: "Sus Scrofa - Compare bisulfite calls vs EMSEQ vs MEGALODON"
author: "P_Terzian"
date: "`r Sys.Date()`"
output:
  prettydoc::html_pretty:
  theme: architect
highlight: github
---

**This report produces a html file as well as a correlation matrix in .tsv format. The number of CpG is printed at the end of the html report.**

```{r, include=FALSE}
library(data.table)
library(Biostrings)
library(ggplot2)
library(DT)
library(reshape2)

```


* To make sure we are always working in the right CG context, we check CG position using the reference genome using the following two chunk of code
* Because megalodon produces 0-based genomic position whereas nfcore/methylseq pos are 1-based, we add 1 to forward strand positions in megalodon results and remove 1 in nfcore results when merging strands.
* Notes : % of coverage for each technologies can be calculated from the number of lines in each tables used to calculate correlation. `nrow` is the function to count the number of line in a table.
* To compute CpG island fraction average, we keep only regions where more than 40% of CpG positions are covered at least by 10 reads. We keep all positions if the region pass this test.


```{r}

getCpGcontextInref <- function(reference, patternToFind = "CG"){
  
  list_CpGTable <- lapply(seq_along(reference), function(x){
  
      currentChrmName <- strsplit(names(reference[x]), " ")[[1]][1]
  
        chrm_pattern_table <- as.data.frame(matchPattern(pattern = patternToFind, subject = reference[[x]])@ranges)
  
      chrm_pattern_table$chrm <-rep(currentChrmName, nrow(chrm_pattern_table))
      
      return(chrm_pattern_table)

  })
AllChrmCpGTable <- do.call(rbind, list_CpGTable)
  

}

```


### Load genome reference
```{r}

sus_scrofa <- readDNAStringSet("/work/project/seqoccin/epigenetique/sus_scrofa/GCA_000003025.6_Sscrofa11.1_genomic.fa")
#chrm1_japonica$`CM003781.1 Coturnix japonica isolate 7356 chromosome 1, whole genome shotgun sequence`[CM003781[strand == "+"]$start]

names_chrm <- sapply(seq_along(sus_scrofa), function(x){
  
      currentChrmName <- strsplit(names(sus_scrofa[x]), " ")[[1]][1]
      
})


sus_scrofa_onlychrm <- sus_scrofa[grep(x = names_chrm, "^CM")]

CG_pos_fullreference <- getCpGcontextInref(sus_scrofa)
setDT(CG_pos_fullreference)
nrow(CG_pos_fullreference)
```


### LOAD megalodon both runs results
```{r}
mega_results_run1 <- fread(file = "/work/project/seqoccin/epigenetique/sus_scrofa/Reference_father/megalodon_calling/sus_scrofa_ref_PAD93746/modified_bases.5mC.bed", stringsAsFactors = FALSE)
colnames(mega_results_run1) <- c("chrm", "start", "stop", "score", "coverage", "strand", "start2", "stop2", "Idonknow", "coverage", "fraction")
hist(mega_results_run1$fraction)
setkey(mega_results_run1, chrm, start)
mega_results_run1 <- mega_results_run1[strand == "+", start := start + 1]
mega_merged1 <- mega_results_run1[, .(coverage = sum(coverage), fraction = sum(fraction) / length(fraction)), by = .(chrm, start)]
setkey(mega_merged1, chrm, start)

mega_results_run2 <- fread(file = "/work/project/seqoccin/epigenetique/sus_scrofa/Reference_father/megalodon_calling/sus_scrofa_ref_PAE20806/modified_bases.5mC.bed", stringsAsFactors = FALSE)
colnames(mega_results_run2) <- c("chrm", "start", "stop", "score", "coverage", "strand", "start2", "stop2", "Idonknow", "coverage", "fraction")
setkey(mega_results_run2, chrm, start)
mega_results_run2 <- mega_results_run2[strand == "+", start := start + 1]
mega_merged2 <- mega_results_run2[, .(coverage = sum(coverage), fraction = sum(fraction) / length(fraction)), by = .(chrm, start)]
setkey(mega_merged2, chrm, start)

megapooled <- rbind(mega_merged1, mega_merged2)
setkey(megapooled, chrm, start)
megapooled <- megapooled[, .(coverage = sum(coverage), fraction = sum(fraction) / length(fraction)), by = .(chrm, start)]

megapooled <- megapooled[coverage < 100 & coverage > 5]

rm(list = c("mega_results_run1", "mega_merged1", "mega_results_run2", "mega_merged2"))

print("this works 1")
```


### Load CpGisland & deal with ONT data
```{r}
cpg_island <- fread(file = "/work/project/seqoccin/epigenetique/sus_scrofa/cpgreport_tsv_sscrofa.tsv", sep="\t")
colnames(cpg_island) <- c("chrm", "start", "stop", "size", "sum", "percentCG", "ObsRatio")
cpg_island$name <- row.names(cpg_island)
setkey(cpg_island, chrm, start, stop)

megapooled$stop <- megapooled$start + 1
setkey(megapooled, chrm, start, stop)

annotated_table_prom <- foverlaps(megapooled, cpg_island, nomatch = 0)

annotated_table_prom <- annotated_table_prom[, count := .N, by = name]

cpgi_freq_prom <- annotated_table_prom[coverage > 9,
                                  .(highconf = .N , freq = mean(fraction), count = unique(count)),
                                  by = name]

cpgi_freq_prom$highconfidence <- ifelse(test =(cpgi_freq_prom$highconf / cpgi_freq_prom$count) >= 0.4, yes = "yes", "no")


rm(megapooled)
```



### Load bismark cytosine_report POOLED
```{r}
pathToFiles = "/work2/project/seqoccin/epigenetique/CelineV/Project_A2P1b-PlaGe-Porc.1470/Pool_run_bisulfite_porc_ref/MethylSeq-nfcore/Sscrofa_per-lane_param-zymo/results/bismark_methylation_calls/stranded_CpG_report/"

list_files <- list.files(path = pathToFiles)

study_DT <- lapply(1:length(list_files), function(x){
    
    
  
    bedcoverageFile           <- fread(file = paste(pathToFiles, list_files[x], sep=""))
    colnames(bedcoverageFile) <- c("chrm", "start", "strand", "Cyt", "Thy", "di_context", "tri_context")
    
    return(bedcoverageFile)
    
  })

study_DT    <- do.call(rbind, study_DT)

study_DT <- study_DT[strand == "-", start := start - 1]

merged_study_DT <- study_DT[, .(Cyt = sum(Cyt), Thy = sum(Thy)), by = .(chrm, start)]

merged_study_DT <- merged_study_DT[Cyt != 0 | Thy != 0]


merged_study_DT[, fraction := Cyt / (Cyt + Thy)]
merged_study_DT[, cov := Cyt + Thy ]

merged_study_DT <- merged_study_DT[cov < 100 & cov >5]

rm(study_DT)
gc()

```


### CpGisland for WGBS 
```{r}

merged_study_DT$stop <- merged_study_DT$start + 1
setkey(merged_study_DT, chrm, start, stop)

annotated_table_BS <- foverlaps(merged_study_DT, cpg_island, nomatch = 0)


annotated_table_BS <- annotated_table_BS[, count := .N, by = name]


cpgi_freq_BS <- annotated_table_BS[cov > 9,
                                   `:=` (highconf = .N , fraction = mean(fraction), count = unique(count)),
                                   by = name]


cpgi_freq_BS[is.na(highconf), highconf := 0]


cpgi_freq_BS$highconfidence <- ifelse(test =(cpgi_freq_BS$highconf / cpgi_freq_BS$count) >= 0.4, yes = "yes", "no")


hist(cpgi_freq_BS[count <100]$count)


cpgi_freq_BS <- cpgi_freq_BS[highconfidence == "yes",
                             .(fraction = mean(fraction), count = unique(count)),
                             by = name]

rm(merged_study_DT)
```

### Load EM-Seq results
```{r}
#Load bismark cytosine_report

pathToFiles = "/work/project/seqoccin/epigenetique/CelineV/Project_A2P1b-PlaGe-Porc.1470/Run_Emseq_porc_ref.18919/MethylSeq-nfcore/Sus_scrofa_param-em-seq/results/bismark_methylation_calls/stranded_CpG_report/"

list_files <- list.files(path = pathToFiles)

study_DT <- lapply(1:length(list_files), function(x){
    
    
  
    bedcoverageFile           <- fread(file = paste(pathToFiles, list_files[x], sep=""))
    colnames(bedcoverageFile) <- c("chrm", "start", "strand", "Cyt", "Thy", "di_context", "tri_context")
    
    return(bedcoverageFile)
    
  })

study_DT    <- do.call(rbind, study_DT)

study_DT <- study_DT[strand == "-", start := start - 1]

merged_study_DT_EMSEQ <- study_DT[, .(Cyt = sum(Cyt), Thy = sum(Thy)), by = .(chrm, start)]

merged_study_DT_EMSEQ <- merged_study_DT_EMSEQ[Cyt != 0 | Thy != 0]

merged_study_DT_EMSEQ[, fraction := Cyt / (Cyt + Thy)]
merged_study_DT_EMSEQ[, cov := Cyt + Thy ]

rm(study_DT)
gc()


##check depth of coverage :
mean(merged_study_DT_EMSEQ$cov)

merged_study_DT_EMSEQ <- merged_study_DT_EMSEQ[cov < 100 & cov > 5]

setkey(merged_study_DT_EMSEQ, chrm, start)

```



### CpGisland for EMseq 
```{r}
merged_study_DT_EMSEQ$stop <- merged_study_DT_EMSEQ$start + 1
setkey(merged_study_DT_EMSEQ, chrm, start, stop)

annotated_table_EMseq <- foverlaps(merged_study_DT_EMSEQ, cpg_island, nomatch = 0)

annotated_table_EMseq <- annotated_table_EMseq[, count := .N, by = name]

cpgi_freq_Emseq <- annotated_table_EMseq[cov > 9,
                                   `:=` (highconf = .N , fraction = mean(fraction), count = unique(count)),
                                   by = name]

cpgi_freq_Emseq[is.na(highconf), highconf := 0]

cpgi_freq_Emseq$highconfidence <- ifelse(test =(cpgi_freq_Emseq$highconf / cpgi_freq_Emseq$count) >= 0.4, yes = "yes", "no")

hist(cpgi_freq_Emseq[count <100]$count)

cpgi_freq_Emseq <- cpgi_freq_Emseq[highconfidence == "yes",
                             .(fraction = mean(fraction), count = unique(count)),
                             by = name]

rm(merged_study_DT_EMSEQ)

```


### Compare megalodon run with themself and BS and Emseq
```{r}

#Megapooled vs BSpooled
setkey(cpgi_freq_prom, name)
setkey(cpgi_freq_BS, name)
setkey(cpgi_freq_Emseq, name)

nrow(cpgi_freq_prom)
nrow(cpgi_freq_BS)
nrow(cpgi_freq_Emseq)


nrow(cpg_island)

megapooledvBSpooled <- cpgi_freq_prom[cpgi_freq_BS, nomatch=0]
nrow(megapooledvBSpooled)
cormegapooledvBSpooled <- cor(megapooledvBSpooled$freq, megapooledvBSpooled$fraction)


megapooledvEMseqpooled <- cpgi_freq_prom[cpgi_freq_Emseq, nomatch=0]
cormegapooledvEMseqpooled <- cor(megapooledvEMseqpooled$freq, megapooledvEMseqpooled$fraction)

rm(mega1vEMseqpooled)
rm(mega2vBSpooled)
rm(megapooledvEMseqpooled)
gc()

```


### Compare WGBS pooled with Emseq
```{r}
EmseqvBSpooled     <- cpgi_freq_Emseq[cpgi_freq_BS, nomatch=0]
nrow(EmseqvBSpooled)
corEmseqvBSpooled<- cor(EmseqvBSpooled$fraction, EmseqvBSpooled$i.fraction)

colpooledwgbs <- c(1, corEmseqvBSpooled, cormegapooledvBSpooled)
colemseq    <- c(  corEmseqvBSpooled, 1, cormegapooledvEMseqpooled)
colmegapooled <- c( cormegapooledvBSpooled, cormegapooledvEMseqpooled,  1)

correlation_matrix <- data.frame(pooledWgbs = colpooledwgbs, Emseq = colemseq, pooledONT = colmegapooled)
rownames(correlation_matrix) <- c("pooledwgbs", "emseq","pooledONT")

fwrite(x = round(correlation_matrix, 2), file = "Correlation_matrix_heatmaptechdatapaper_cpgislands_average.tsv", sep = "\t")

```
