---
title: "Compare bisulfite calls vs EMSEQ vs MEGALODON"
author: "P_Terzian"
date: "`r Sys.Date()`"
output:
  prettydoc::html_pretty:
  theme: architect
highlight: github
---


```{r, include=FALSE}
library(data.table)
library(Biostrings)
library(ggplot2)
library(DT)
library(reshape2)
library(ggridges)

```

```{r}

getCpGcontextInref <- function(reference, patternToFind = "CG"){
  
  list_CpGTable <- lapply(seq_along(reference), function(x){
  
      currentChrmName <- strsplit(names(reference[x]), " ")[[1]][1]
  
        chrm_pattern_table <- as.data.frame(matchPattern(pattern = patternToFind, subject = reference[[x]])@ranges)
  
      chrm_pattern_table$chrm <-rep(currentChrmName, nrow(chrm_pattern_table))
      
      return(chrm_pattern_table)

  })
AllChrmCpGTable <- do.call(rbind, list_CpGTable)
  

}

```


### Load genome reference
```{r}

full_japonica <- readDNAStringSet("/work/project/seqoccin/data/species/coturnix_japonica/Coturnix_japonica_2.0/GCA_001577835.1_Coturnix_japonica_2.0_genomic.fa")
#chrm1_japonica$`CM003781.1 Coturnix japonica isolate 7356 chromosome 1, whole genome shotgun sequence`[CM003781[strand == "+"]$start]

names_chrm <- sapply(seq_along(full_japonica), function(x){
  
      currentChrmName <- strsplit(names(full_japonica[x]), " ")[[1]][1]
      
})


full_japonica_onlychrm <- full_japonica[grep(x = names_chrm, "^CM")]

CG_pos_fullreference <- getCpGcontextInref(full_japonica)
setDT(CG_pos_fullreference)
setkey(CG_pos_fullreference, chrm, start)

### Compute numbers of sites per chromosome
nb_CG_per_chrm <- as.data.table(table(CG_pos_fullreference$chrm))

colnames(nb_CG_per_chrm) <- c("chrm", "position")

nb_CG_per_chrm$type <- "reference"

nbCGref <- nrow(CG_pos_fullreference)
```


# LOAD megalodon both runs results
```{r}
mega_results_run1 <- fread(file = "/work/project/seqoccin/epigenetique/coturnix_japonica/PromethION_PAE47187_Promethion_PAE47965_natifs/megalodon_calling/2022_last_calls_with_bam_fastq_mod/coturnix_ref_PAE47187/modified_bases.5mC.bed", stringsAsFactors = FALSE)
colnames(mega_results_run1) <- c("chrm", "start", "stop", "score", "count", "strand", "start2", "stop2", "Idonknow", "count", "fraction")
# hist(mega_results_run1$fraction)
setkey(mega_results_run1, chrm, start)
mega_results_run1 <- mega_results_run1[strand == "+", start := start + 1]
mega_merged1 <- mega_results_run1[, .(count = sum(count), fraction = sum(fraction) / length(fraction)), by = .(chrm, start)]
setkey(mega_merged1, chrm, start)



mega_results_run2 <- fread(file = "/work/project/seqoccin/epigenetique/coturnix_japonica/PromethION_PAE47187_Promethion_PAE47965_natifs/megalodon_calling/2022_last_calls_with_bam_fastq_mod/coturnix_ref_PAE47965/modified_bases.5mC.bed", stringsAsFactors = FALSE)
colnames(mega_results_run2) <- c("chrm", "start", "stop", "score", "count", "strand", "start2", "stop2", "Idonknow", "count", "fraction")
setkey(mega_results_run2, chrm, start)
mega_results_run2 <- mega_results_run2[strand == "+", start := start + 1]
mega_merged2 <- mega_results_run2[, .(count = sum(count), fraction = sum(fraction) / length(fraction)), by = .(chrm, start)]
setkey(mega_merged2, chrm, start)



megapull <- rbind(mega_merged1, mega_merged2)
setkey(megapull, chrm, start)
megapull <- megapull[, .(count = sum(count), fraction = sum(fraction) / length(fraction)), by = .(chrm, start)]

megapull_uncomplete <- megapull[CG_pos_fullreference, all = T]
megapull_uncomplete[is.na(count), count := 0]

rm(mega_merged1)
rm(mega_merged2)
gc()

```

# LOAD Pacbio run1 & 2 result
```{r}
primrose_results_run1 <- fread(file = "/work/project/seqoccin/epigenetique/coturnix_japonica/PacBio/reference_coturnix/analysis/run_200820_203625/m64122_200820_203625.ccs.bedmethyl", stringsAsFactors = FALSE)
colnames(primrose_results_run1) <- c("chrm", "start", "stop", "name", "score", "strand", "start2", "stop2", "Idonknow", "count", "fraction", "cannonical", "modified", "filtered")
hist(primrose_results_run1$fraction)
setkey(primrose_results_run1, chrm, start)
primrose_results_run1 <- primrose_results_run1[strand == "+", start := start + 1]
primrose_merged1 <- primrose_results_run1[, .(count = sum(count), fraction = sum(fraction) / length(fraction)), by = .(chrm, start)]
setkey(primrose_merged1, chrm, start)


primrose_results_run2 <- fread(file = "/work/project/seqoccin/epigenetique/coturnix_japonica/PacBio/reference_coturnix/analysis/run_201001_030231/m64122_201001_030231.ccs.bedmethyl", stringsAsFactors = FALSE)
colnames(primrose_results_run2) <- c("chrm", "start", "stop", "name", "score", "strand", "start2", "stop2", "Idonknow", "count", "fraction", "cannonical", "modified", "filtered")
hist(primrose_results_run2$fraction)
setkey(primrose_results_run2, chrm, start)
primrose_results_run2 <- primrose_results_run2[strand == "+", start := start + 1]
primrose_merged2 <- primrose_results_run2[, .(count = sum(count), fraction = sum(fraction) / length(fraction)), by = .(chrm, start)]
setkey(primrose_merged2, chrm, start)

primrose_merged <- rbind(primrose_merged1, primrose_merged2)
setkey(primrose_merged, chrm, start)
primrose_merged <- primrose_merged[, .(count = sum(count), fraction = sum(fraction) / length(fraction)), by = .(chrm, start)]


primrose_uncomplete <- primrose_merged[CG_pos_fullreference, all = T]
primrose_uncomplete[is.na(count), count := 0]


nrow(primrose_uncomplete)


rm(list = c("primrose_results_run1", "primrose_merged1", "primrose_results_run2", "primrose_merged2"))


```


## Load bismark cytosine_report PULL
```{r}
pathToFiles = "/work/project/seqoccin/epigenetique/CelineV/Project_A2P1a-PlaGe-Caille.1447/Pool_run_bisulfite/MethylSeq-nfcore/coturnix_per-lane_param-zymo/results/bismark_methylation_calls/stranded_CpG_report/"

list_files <- list.files(path = pathToFiles)

study_DT <- lapply(1:length(list_files), function(x){
    
    
  
    bedcoverageFile           <- fread(file = paste(pathToFiles, list_files[x], sep=""))
    colnames(bedcoverageFile) <- c("chrm", "start", "strand", "Cyt", "Thy", "di_context", "tri_context")
    
    return(bedcoverageFile)
    
  })

study_DT    <- do.call(rbind, study_DT)

merged_study_DT_stranded <- study_DT[, .(Cyt = sum(Cyt), Thy = sum(Thy)), by = .(chrm, strand, start)]

study_DT <- study_DT[strand == "-", start := start - 1]

merged_study_DT_stranded[, fraction := Cyt / (Cyt + Thy)]
merged_study_DT_stranded[, cov := Cyt + Thy ]


merged_study_DT <- study_DT[, .(Cyt = sum(Cyt), Thy = sum(Thy)), by = .(chrm, start)]

merged_study_DT[, fraction := Cyt / (Cyt + Thy)]
merged_study_DT[, cov := Cyt + Thy ]

rm(study_DT)
rm(merged_study_DT_stranded)
gc()

```


### Load EM-Seq results
```{r}
# Load bismark cytosine_report

pathToFiles = "/work/project/seqoccin/epigenetique/CelineV/Project_A2P1a-PlaGe-Caille.1447/Run_Emseq_caille_ref.18827/MethylSeq-nfcore/Coturnix_japonica_other-param-emseq/results/bismark_methylation_calls/stranded_CpG_report/"

list_files <- list.files(path = pathToFiles)

study_DT <- lapply(1:length(list_files), function(x){
    
    
  
    bedcoverageFile           <- fread(file = paste(pathToFiles, list_files[x], sep=""))
    colnames(bedcoverageFile) <- c("chrm", "start", "strand", "Cyt", "Thy", "di_context", "tri_context")
    
    return(bedcoverageFile)
    
  })

study_DT    <- do.call(rbind, study_DT)

merged_study_DT_stranded <- study_DT[, .(Cyt = sum(Cyt), Thy = sum(Thy)), by = .(chrm, strand, start)]


#merged_study_DT_stranded <- merged_study_DT_stranded[Cyt != 0 | Thy != 0]

study_DT <- study_DT[strand == "-", start := start - 1]

merged_study_DT_stranded[, fraction := Cyt / (Cyt + Thy)]
merged_study_DT_stranded[, cov := Cyt + Thy ]


merged_study_DT_EMSEQ <- study_DT[, .(Cyt = sum(Cyt), Thy = sum(Thy)), by = .(chrm, start)]

#merged_study_DT_EMSEQ <- merged_study_DT_EMSEQ[Cyt != 0 | Thy != 0]


merged_study_DT_EMSEQ[, fraction := Cyt / (Cyt + Thy)]
merged_study_DT_EMSEQ[, cov := Cyt + Thy ]


rm(study_DT)
gc()
rm(merged_study_DT_stranded)
gc()


```


### Compare megalodon run with themself and BS and Emseq
```{r}

list_cov_all <- lapply(seq(0,100,5), function(x){

      nbBS <- round(nrow(merged_study_DT[cov > x]) / nbCGref * 100)
      nbEmseq <- round(nrow(merged_study_DT_EMSEQ[cov > x]) / nbCGref * 100)
      nbONT <- round(nrow(megapull_uncomplete[count > x]) / nbCGref * 100)
      nbPB <- round(nrow(primrose_uncomplete[count > x]) / nbCGref * 100)
      filtre <- x     
      return(c(nbPB,nbBS, nbEmseq, nbONT, filtre))
})

# list_cov_all <- readRDS(file = "list_nbpos_at_differentfilter_pacbio.RDS")
dt_nbposcovered <- rbindlist(lapply(list_cov_all, as.data.frame.list))
colnames(dt_nbposcovered) <- c("PacBio","WGBS", "EM-Seq", "ONT", "filtre")



dt_nbposcovered <- melt(dt_nbposcovered, measure.vars = c("ONT", "EM-Seq", "WGBS", "PacBio"))
 
ggplot(data=dt_nbposcovered, aes(x=filtre, y=value, group=variable, color=variable, fill=variable)) +
   geom_density_line(stat = "identity", size=.5, alpha=0.2)+
   labs(title = "CpG site coverage at different minimum depths", expand=c(0,0)) +
   scale_y_continuous(name="percentage of CpGs (%)") +
   scale_x_continuous(name="min depth of coverage (X)", expand=c(0,0), breaks=seq(0,100,10)) +
   theme_bw()

ggsave("DTpaper_cumulative_coverage_coturnix.pdf")

```


